### PR TITLE
Devel

### PR DESCRIPTION
**Pull request by kaidokert**
## Summary by CodeRabbit

- **New Features**
  - Introduced new adjacency list, edge list, and matrix-based graph structures with improved trait-based APIs for flexible graph representations.
  - Added new graph algorithms including connected components and Tarjan's strongly connected components.
  - Added extensive new example programs demonstrating various graph algorithms such as Bellman-Ford, Dijkstra, BFS, DFS (recursive and iterative), greedy coloring, Kruskal's MST, topological sort, and topological sort comparison.
  - Added support for stack-friendly, no_std-compatible data structures, including probing-based sets, maps, and circular queues.
  - Added new probing module for shared linear probing logic used in maps and sets.

- **Bug Fixes**
  - Improved collision handling and deletion in sets and maps via linear probing and tombstone logic.

- **Refactor**
  - Split and modernized graph trait interfaces into reference-based and value-based variants.
  - Simplified and unified error handling across graph and algorithm modules.
  - Reorganized and modularized codebase for clarity, extensibility, and maintainability.
  - Refactored algorithms for improved clarity, consistency, and trait usage.
  - Updated traversal APIs to use references for start nodes and refined error handling.
  - Replaced previous edge list implementations with new modular edgelist modules.
  - Updated container implementations to use option-wrapped buffers supporting non-Copy types.
  - Removed deprecated traits and consolidated related traits for edges and nodes.
  - Enhanced error enums and streamlined algorithm error variants.
  - Updated build and dependency configurations removing `num-traits` and adding `test-log`.
  - Improved Kahn's algorithm to use BFS semantics and simplified cycle detection.
  - Refined Kruskal's algorithm with clearer union-find logic and better error handling.
  - Updated Bellman-Ford and Dijkstra algorithms to use optional distance maps and improved trait bounds.
  - Redesigned greedy coloring algorithm with comprehensive documentation and expanded tests.
  - Modularized adjacency list implementation into submodules with validation and trait implementations.
  - Updated circular queue to use option-wrapped storage supporting non-Copy types.
  - Enhanced adjacency matrix implementations with flexible node indexing and efficient bit-packed storage.

- **Documentation**
  - Expanded and clarified documentation throughout the codebase and examples for easier adoption and understanding.
  - Added detailed algorithm and data structure documentation for new and refactored modules.
  - Updated crate-level documentation and example usage to reflect API changes.
  - Improved documentation comments for graph traits, node and edge traits, and visitation tracking.

- **Tests**
  - Added comprehensive tests for all new data structures, algorithms, and edge cases.
  - Added extensive collision and tombstone reuse tests for maps and sets.
  - Rewrote and expanded tests for traversal algorithms and graph structures using concrete graph types.
  - Added tests for new adjacency list, edge list, and matrix graph implementations.
  - Expanded tests for algorithms including connected components, Tarjan's SCC, greedy coloring, Kruskal's MST, and topological sorts.
  - Added tests verifying correctness of matrix-based graphs, probing logic, and circular queue behavior.
  - Improved test coverage for error handling and edge cases across all modules.

### CODE REVIEW

File: src/containers/maps/staticdict.rs [hunk 1]
```diff
@@ -130,17 +193,33 @@ where
...
     fn clear(&mut self) {
-        self.items.fill_with(|| None);
+        self.slots = core::array::from_fn(|_| Slot::Empty);
     }
 }
 
 #[cfg(test)]
 mod test {
     use super::*;
+    use crate::tests::array_collect;
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/containers/maps/staticdict.rs:203]: It's actually in `tests` module at the toplevel `lib.rs` module

File: src/matrix/simple_matrix.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+                    .get(node)?
+                    .as_ref()
+                    .get(col_index)?
+                    .as_ref()
+                    .map(|_| col_index)
+            } else {
+                None
+            }
+        }))
+    }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/matrix/simple_matrix.rs:75]: The key issue here ( and elsewhere in Matrix impls ) is that this function can't have two return paths - it would produce a iterator type mismatch on return value. Because the return value is `impl Iterator` it can only resolve to a single concrete type

File: src/matrix/map_matrix/by_ref.rs [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+            .outgoing_edges(matrix_idx.unwrap_or(usize::MAX))
+            .unwrap()
+            .filter(move |_| matrix_idx.is_some()) // Filter out everything if node doesn't exist
+            .filter_map(move |target_idx| {
+                self.index_map
+                    .iter()
+                    .find(|(_, &idx)| idx == target_idx)
+                    .map(|(node, _)| node)
+            }))
+    }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/matrix/map_matrix/by_ref.rs]: Can you re-check this ? Last commit removed .unwrap. But again - as above, can't have two return paths, which is why there's a two-stage filter here.

File: src/graph.rs [hunk 4]
```diff
@@ -2,184 +2,258 @@
...
+        node: &'a NodeIndex,
+    ) -> Result<impl Iterator<Item = &'a NodeIndex>, Self::Error>
+    where
+        NodeIndex: 'a,
+    {
+        Ok(self
+            .iter_edges()?
+            .filter(move |(src, _dst)| *src == node)
+            .map(|(_src, dst)| dst))
     }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/graph.rs:68]: Done that for adjacency lists, but for edgelist and edgenodelist there's no better outgoing_edges method available.

File: src/algorithms/tarjan_scc.rs [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+                &mut node_buffer[buffer_offset..],
+                &mut component_sizes[component_count..],
+            )?;
+            component_count += new_components;
+            buffer_offset += new_buffer_offset;
+
+            if component_count >= components.len() {
+                return Err(AlgorithmError::ResultCapacityExceeded);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/algorithms/tarjan_scc.rs]: I think this is now addressed

File: src/algorithms.rs [hunk 6]
```diff
@@ -7,72 +7,63 @@
...
-#[cfg(feature = "num-traits")]
+mod connected_components;
 mod dijkstra;
-#[cfg(feature = "heapless")]
 mod greedy_coloring;
 mod kahns;
 mod kruskals;
+mod tarjan_scc;
 mod topological_sort;
 mod traversal;
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/algorithms.rs:18]: This seems to be fixed

File: src/adjacency_list/outgoing_nodes.rs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    E: NodesIterable<Node = NI>,
+{
+    type Iter<'a>: DoubleEndedIterator<Item = &'a NI>
+    where
+        NI: 'a,
+        Self: 'a;
+    fn as_outgoing_nodes<'a>(&'a self) -> Self::Iter<'a>
+    where
+        NI: 'a;
+}
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/adjacency_list/outgoing_nodes.rs:17]: I think the E argument is there for automatically binding NodesIterable constraint in all dependents. Trying to remove breaks all compilation and forces specific types to be provided at construction time for everything
